### ISSUE TITLE
Changelog

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback-next#263]
We should start writing changelog to accompany each new release.

Depending on the outcome of #262, we may want this changelog to be per-package (e.g. `@loopback/testlab` vs `@loopback/context`) or a single global one.

In my experience, auto-generated changelog (as `slt-release` creates for us) has limited usability for package consumers because it's too brief.

I am proposing to start writing Changelog entries as part of Pull Requests, require each Pull Request to include a descriptive Changelog entry that package users can read to understand the changes made.

@ritch @raymondfeng @superkhau

**Comment by superkhau** [comment]
I'm on the fence -- I spoke to @rmg briefly regarding https://www.npmjs.com/package/standard-version for continuously deployment possibilities (basically you have to use a formatted git commit message which allows for auto deploying + changelog generation). @ritch wanted continuous deployment over push button deploy (discussions still up in the air). Anyways, some decisions we need to make:

1. Push button deploy vs continuous deployment
2. If continuous deployment is chosen, we can format our messages and get free changelog generation (whether it's detailed enough for our intents needs a spike)
3. If push button deploy, we'll need to manually add changelog entries (or find another lib to help us figure out the generation from the PR messages/git description you're proposing).

Either way, I don't think it's at the top of our backlog ATM, but definitely something that should be spiked on. Issue to follow at https://github.com/strongloop/loopback-next/issues/194.

**Comment by bajtos** [comment]
I slightly prefer push-button deploy over continuous deployment, because it allows us to land multiple incremental changes and publish a version only after the full feature is finished. With continuous deployment, I think we will need to use feature flags more often, in order to hide work that has landed on master (landed because we want continuous integration) but that's not complete and ready for public consumption yet (landed because we want to quickly iterate in small increments). The alternative to feature flags is long-lived feature branches, but I hope we all agree that's a no-go.

As for Changelog generation: in my experience, auto-generated changelogs are rarely useful. A good commit message requires different information than a good changelog entry. The commit message is for future framework developers, therefore it often explains internal implementation details and references to other related code/concepts. OTOH, the changelog entry is for framework users, therefore it should explain the impact on existing applications, upgrade path, etc.

Monorepo is another point to consider. There is single git history for all packages, but each package should have its own Changelog that makes sense in the context of that package. A single commit touching several places may require several different Changelog entries, each entry focused on a different level of abstraction, depending on which package it is in.

**Comment by superkhau** [comment]
I'm +1 for push button deploy for more control. Too many edge cases to work out with continuous deployment (or both if we want to only do it for simple repos like @loopback/testlab). Continuous deploy is also dangerous in that non-working versions may actually/continually go out, whereas we probably at least smoke tested the push button deploys.

Sounds like changelog is up for debate. I think we should spike to gather all the things expected from a changelog and get some community feedback. Include those things and see if we can automate the final decision.

+1 on Monorepo concerns, I'm starting to not like the monorepo approach as much beause it feels so against what is typically isolated modules (along with all the changelogs, history, issues, etc). Sometimes, it feels like if we just made a bunch of subdirs in `loopback-next`, we would get all the same benefits without all the tooling configs/deployment/config overhead. Literally:

```
index.js
lib
src
- context
- openapi-spec
- openapi-spec-builder
- testlab
```

without independent versioning lol. Just release things in batches (exactly like how git history shows) and log them all in one changelog. Anyways, flame war topic lol, so I'll stop here.

**Comment by kjdelisle** [comment]
This might help: https://github.com/lerna/lerna-changelog

**Comment by kjdelisle** [comment]
Would still require some manual reviewing to ensure that the correct package names are being used in commit messages, but it should help.

**Comment by bajtos** [comment]
The scope for Sprint 42: research what are the existing tools for generating packages-scoped changelog entries. Talk to @kjdelisle @superkhau @bajtos to decide which one to use going forward. Stretch goal: implement this new changelog generation.

**Comment by virkt25** [comment]
# SPIKE Results

## Problem
Need a way to be able to generate `CHANGELOG.MD` for each individual package in the monorepo when a new version is published. Should be ideally automated.

## Potential Solutions
A few different approaches were investigated and tested. They are discussed below along with the pros and cons for each as well as my personal recommendation for the approach we should take based on the result but that's up for discussion!

### 1. lerna-changelog
- Third party npm module for generating a changelog for lerna monorepos. Build a changelog based on git commits and the configured labels that should be added to PR's. 

#### Usage
- Run `lerna-changelog` and copy & paste into `CHANGELOG.MD`

#### Pros
- Headings and labels can be customized through a config option
- PR's are linked for reference in the output. 

#### Cons
- Requires us to make sure all PR's have the appropriate label on it otherwise it'll be ignored from the logs. 
- Output format cannot be customized
- It creates only one `CHANGELOG.MD` for the entire monorepo. 
  - But each commit lists the packages that are affected so we can in theory parse the information and have a script which runs the tool, converts it into a Changelog for each package and commit that. 
  - We can fork it to make it work as well if needed
- Requires a GITHUB Auth Token to work
- Have to copy paste the Changelog Manually (But can be automated)

#### Sample Output
```
## Unreleased (2017-08-18)

#### Feature
* `authentication`, `core`
  * [#492](https://github.com/strongloop/loopback-next/pull/492) authentication: make it a proper component + cleanup. ([@bajtos](https://github.com/bajtos))
  * [#444](https://github.com/strongloop/loopback-next/pull/444) Rename sequence actions to start with "sequence.actions." prefix. ([@bajtos](https://github.com/bajtos))
* `context`
  * [#492](https://github.com/strongloop/loopback-next/pull/492) authentication: make it a proper component + cleanup. ([@bajtos](https://github.com/bajtos))
* `core`, `example-codehub`, `openapi-spec-builder`, `openapi-spec`, `testlab`
  * [#475](https://github.com/strongloop/loopback-next/pull/475) Expose "GET /openapi.json", implement spec validation. ([@bajtos](https://github.com/bajtos))
* `core`, `openapi-spec`
  * [#460](https://github.com/strongloop/loopback-next/pull/460) Add more `@param.*` shortcuts. ([@bajtos](https://github.com/bajtos))
* `core`
  * [#459](https://github.com/strongloop/loopback-next/pull/459) Add shortcut decorators for HTTP verbs. ([@bajtos](https://github.com/bajtos))
  * [#457](https://github.com/strongloop/loopback-next/pull/457) Add `@param` decorator + shortcuts. ([@bajtos](https://github.com/bajtos))
  * [#445](https://github.com/strongloop/loopback-next/pull/445) core: support @get and @operation decorators. ([@bajtos](https://github.com/bajtos))
* `core`, `example-codehub`
  * [#443](https://github.com/strongloop/loopback-next/pull/443) Add app.route() API for controller-based routes. ([@bajtos](https://github.com/bajtos))

#### Committers: 8
- Deepak Rajamohan ([deepakrkris](https://github.com/deepakrkris))
- Loay ([loay](https://github.com/loay))
- Mahesh Patsute ([mpatsute](https://github.com/mpatsute))
- Miroslav Bajtoš ([bajtos](https://github.com/bajtos))
- Raymond Feng ([raymondfeng](https://github.com/raymondfeng))
- Sakib Hasan ([ssh24](https://github.com/ssh24))
- Taranveer Virk ([virkt25](https://github.com/virkt25))
- Tetsuo Seto ([Setogit](https://github.com/Setogit))
```

### 2. --conventional-commits
- Built right into lerna publish. Automatically creates a `CHANGELOG.MD` for each package based on commit message history as well as automatic versioning. Require commits to follow the [conventional-commit standard version commit format](https://github.com/conventional-changelog/standard-version) which is based on [angular commit convention](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit)

#### Usage
`lerna publish --conventional-commits` 
For testing:
`lerna publish --skip-npm --skip-git --conventional-commits`

#### Pros
- Automatically updates package.json version
- Update CHANGELOG for each changed package (since last release)
- Automatic tagging.
- Commit format can be enforced as `Required` by CI

#### Cons
- Requires us to use the Angular Convention (can be a pro. It's simple and intuitive.) We would have to update out `CONTRIBUTING` guide and it's not our style. 
  - Example

```
<type>(<scope>): <subject>
<BLANK LINE>
<body>
<BLANK LINE>
<footer>
```

- Type is one of: ```build, ci, docs, feat, fix, perf, refactor, style, test```
- Scope is a package name
- 100 characters for the line (vs. 50 for us). 
- No flexibility in changing the commit format at the moment. (We can open an issue to allow support for custom commit formats. The underlying package supports it but lerna has only intergerated the angular convention.

#### Sample Output
- No commits matching the format at the moment in any of the packages. The default that got generated is nice. 
```
# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="4.0.0"></a>
# [4.0.0](https://github.com/strongloop/loopback-next/compare/@loopback/openapi-spec-builder@4.0.0-alpha.4...@loopback/openapi-spec-builder@4.0.0) (2017-08-21)
```
- Here are some other linked examples of an actual changelog that would get generated. The `conventional-changelog` project itself is a `lerna` repo. [Conventional Changelog monorepo](https://github.com/conventional-changelog/conventional-changelog)
  - [conventional-changelog-angular](https://github.com/conventional-changelog/conventional-changelog/blob/master/packages/conventional-changelog-angular/CHANGELOG.md)
  - [conventional-changelog-cli](https://github.com/conventional-changelog/conventional-changelog/blob/master/packages/conventional-changelog-cli/CHANGELOG.md)

### 3. Custom Scripts
- We can write a custom script for going through github and creating a changelog. 

#### Pros
- We have full control over output format, etc. We could even fork `lerna-changelog` if we wanted to since it lays the foundation pretty well. 

#### Cons
- Another package / module to write and maintain

#### Sample Output
- TBD. We have full control. 

### 4. Manual
- We can make it so that each PR must be accompanied by an updated entry in the `CHANGELOG.md`

#### Pros
- Can be enforced by CI

#### Cons
- Easy to overlook and forget at first (will remember when enforced)
- It's manual. Hard to track if something should've be a major / minor version bump. 

## Recommendation
My vote is for option #2 since it's an open source project, has a reasonable format for us to follow, auto bumps versions, can be enforced by a Commit Message Linter, generates a changelog for each modified package. 

cc: @bajtos @superkhau @kjdelisle

**Comment by bajtos** [comment]
Thank you for the detailed summary, @virkt25. I agree with your recommendation to use conventional commits together with their built-in lerna support.

Because conventional commits is a new concept that we are not used to, I think it's crucial to have a good commit message linter configured in our CI. We already have some commit linting configured, please talk to @rmg and find out whether and how it can be customized to enforce conventional commits.

**Comment by virkt25** [comment]
I had a chance to talk with @rmg and discuss CI linting possibilities. The CI Environment that does the linting check currently doesn't have Node . It uses a series of `Ruby` files controlled by Groovy. 

That said it would most certain be possible for us to translate the rules used by the `conventional-changelog` linter to our own script and then configure the CI to use that against the `loopback-next` Repository only since the new format would only apply here (for now). 

#### References: 
comventional-commit/validate-commit-msg Validator we would have to port to Ruby. 
https://github.com/conventional-changelog/validate-commit-msg/blob/master/lib/validateMessage.js

**Comment by bajtos** [comment]
Alternatively, can we perhaps disable our Ruby-powered commit-message linter (e.g. by a flag in `package.json`) and then implement commit-message linter as part of our regular build which runs `npm install && npm test` and therefore have Node.js available?

Upsides:
 - We (LoopBack developers) have full control over our commit linting
 - Developers can install local pre-commit git hook

Downsides:

 - We will be running commit-message linter multiple times, basically on every CI machine.
 - Our custom linter needs to understand which commits are a part of the pull request, this may be different for Travis, AppVeyor and Jenkins. Extra work is required.

**Comment by rmg** [comment]
The PR/commit linting isn't based on package.json or any other repo content. The repo scanner for that automation only cares whether a repo is public or private to determine whether the job goes on a public or private Jenkins. Disabling it or configuring an alternate scanner for a repo would require more or less the same modification of the scanner.

The only thing that the current linter would need to change in order to not give false-negatives is the line widths; the other aspects of commit message formatting _could_ be done separately as part of `npm test`.

**Comment by kjdelisle** [comment]
@rmg @bajtos Is there any reason we can't switch our commit linting rules org-wide to use the new rules? AFAIK, the slt-release tool wouldn't be affected by this change, and then we can just change the appropriate job step to use the new linting rules. It would be a bit abrupt for other modules, but it would definitely be easier than phasing in an opt-out flag and then having to phase it back out in the future once loopback-next is the new way (and the old repos get marked obsolete and taken out of CI).

**Comment by rmg** [comment]
I like that it is a structured format, but I really dislike the maximum column width being doubled.

That said, it is entirely possible that I'm the only one who actually cares about.

**Comment by bajtos** [comment]
> I really dislike the maximum column width being doubled

Yeah, I am not very happy about that either. But I'd rather double the column width than maintain yet another changelog generator.

**Comment by Nosfistis** [comment]
> Another way to use https://github.com/strongloop/loopback-next/commits/master/packages/cli/package.json and click on `...` for each commits with `chore: publish release`.

That makes it possible, but I don't think a commit view is the best way. I find it better for a developer to check any notable changes in the framework if the changelogs were concatenated into a central one.
